### PR TITLE
properly unlock flash on re-entry of wolfBoot_swap_and_final_erase

### PR DESCRIPTION
**Pull request by jpbland1**
and move it out of the lock logic of update and delta update

### CODE REVIEW

**Comment by jpbland1** [comment]
added hal and ext flash locks to the simulator, fixed test failures caused by flash being accessed while locked. @danielinux any idea why the unit test is failing? is it because I added the locks to hal/sim.c and it's not aware of them?

**Comment by danielinux** [comment]
The hal_flash_lock/unlock functions are mocked and checked, it seems that the unit test is dectecting a double lock.

**Comment by jpbland1** [comment]
I figured out the issue, it was only unlocking flash when encryption was on, I think it was a copy and paste issue. should be good to go now

File: src/update_flash.c [hunk 1]
```diff
@@ -480,15 +494,17 @@ static int wolfBoot_delta_update(struct wolfBoot_image *boot,
...
 out:
 #ifdef EXT_FLASH
     ext_flash_lock();
 #endif
     hal_flash_lock();
+    /* start re-entrant final erase, return code is only for resumption in
+     * wolfBoot_start */
+    if (ret == 0) {
+        wolfBoot_swap_and_final_erase(0);
+    }
```
Comment (danielinux) [inline src/update_flash.c:507]: I think this should still be under `#ifndef DISABLE_BACKUP`, unless it's needed anyway. If that's the case, remove the empty `#ifndef` clause at L 496 then.

**Review by danielinux — CHANGES_REQUESTED** [review]
Looks good. Good idea adding checks for unlock/lock in the simulator as well.

Just a doubt about `DISABLE_BACKUP` calling `wolfBoot_swap_and_final_erase()`

File: src/update_flash.c (hunk 1 above)
Comment (jpbland1) [reply src/update_flash.c]: missed that, fixed now

**Review by danielinux — APPROVED** [review]

**Pull request merged**